File: jianzhioffer/firstround/jianzhi_60.py
class Solution(object):
    def dicesProbability(self, n):
        """
        :type n: int
        :rtype: List[float]
        """
        dp = [0] + [1]*6  # 初始状态，只有一个骰子时，数字和为1-6，且各个数字只出现一次，为了让数组下标和实际情况吻合，故令dp[0]=0，从dp[1]开始计算

        for i in range(2, n+1):
            dp += [0]*6
            for j in range(i*6, i-1, -1):  # n个骰子的最大数字和为6n，最小数字和为n。比如n=2时，显然可知2个骰子的最大数字和为12，最小数字和为2。
                dp[j] = 0  # 注意要令dp[j]=0，因为不考虑n-1个骰子的结果。比如n=2时，那么要先令dp[6]=0，即不考虑原本n=1时骰子和为6的情况了。
                for cur in range(j-1, j-7, -1):
                    if cur < i-1:
                        break
                    dp[j] += dp[cur]
        # 最后返回结果也是取dp[n:]，因为n个骰子的最小数字和为n，所以前面dp[:n]的值舍弃掉
        return [x / pow(6, n) for x in list(map(float, dp[n:]))]


if __name__ == '__main__':
    obj = Solution()
    result = obj.dicesProbability(2)
    print(result)

File: jianzhioffer/secondround/jianzhi_60.cpp
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;

class Solution
{
public:
	vector<double> dicesProbability(int n);
};	


//动态规划，时间复杂度O(N^2)，空间复杂度O(N^2)
//vector<double> Solution::dicesProbability(int n)
//{
//	if(n<1) return vector<double>{};
//	vector<double> res {};
//	vector<vector<int>> dp(n+1, vector<int>(6*n+1,0));
//	for(int i=1; i<=6; i++)
//	{
//		dp[1][i]=1;
//	}
//
//	for(int i=2; i<=n; i++)
//	{
//		for(int j=i; j<=6*i; j++)
//		{
//			for(int k=1; k<=6; k++)
//			{
//				if(j-k>0)
//				{
//					dp[i][j] += dp[i-1][j-k];
//				}
//				else break;
//			}
//		}
//	}
//	
//	double count = pow(6.0, n);
//	for(int i=n; i<=6*n; i++)
//	{
//		res.push_back(dp[n][i]/count);
//	}
//	return res;
//}

//优化空间的动态规划。不建二维数组dp，dp只用一维，另建一个辅助数组存储n-1个骰子各点数出现的次数
//时间复杂度O(N^2)，空间复杂度O(N)
vector<double> Solution::dicesProbability(int n)
{
	if(n<1) return vector<double>{};
	vector<double> res {};
	vector<int> tmp(6*n+1,0);
	
	for(int i=1; i<=6; i++) tmp[i] = 1;
	
	for(int i=2; i<=n; i++)
	{
		vector<int> dp(6*i+1,0);
		for(int j=i; j<=6*i; j++)
		{
			for(int k=1; k<=6; k++)
			{
				if(j-k>0) dp[j] += tmp[j-k];
				else break; 
			}
		}

		copy(dp.begin(), dp.end(), tmp.begin());
	}

	double counts = pow(6, n);
	for(int i=n; i<=6*n; i++)
	{
		res.push_back(tmp[i]/counts);
	}
	return res;
}

int main()
{
	Solution obj;
	vector<double> result = obj.dicesProbability(2);
	return 0;
}
